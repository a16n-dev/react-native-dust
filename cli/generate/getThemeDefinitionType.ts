import { debug } from "../debug";

const DEFAULT_THEME_KEYS = ["colors", "spacing", "radius", "shadow"];

/**
 * This takes in a theme object and generates a TypeScript type to account for any
 * custom properties the user may have defined in their theme.
 */
export async function getThemeDefinition(
  themes: Record<string, any>,
): Promise<string> {
  // Get custom properties (not in default DuckTheme)
  const themeNames = Object.keys(themes);
  const firstTheme = themes[themeNames[0]];

  // Extract only custom properties
  const customProperties = Object.keys(firstTheme).filter(
    (key) => !DEFAULT_THEME_KEYS.includes(key),
  );

  debug("Theme has custom properties:", customProperties);

  // Generate type definitions for custom properties only
  function generateCustomPropertyType(value: any): string {
    if (typeof value === "string") return "string";
    if (typeof value === "number") return "number";
    if (typeof value === "boolean") return "boolean";
    if (Array.isArray(value)) {
      if (value.length === 0) return "any[]";
      const firstType = typeof value[0];
      return value.every((item) => typeof item === firstType)
        ? `${firstType}[]`
        : "any[]";
    }
    if (typeof value === "object" && value !== null) {
      const entries = Object.keys(value).map((key) => {
        const propType = generateCustomPropertyType(value[key]);
        return `    ${key}: ${propType};`;
      });
      return `{\n${entries.join("\n")}\n  }`;
    }
    return "any";
  }

  const customTypeEntries = customProperties.map((key) => {
    const value = firstTheme[key];
    const type = generateCustomPropertyType(value);
    return `  ${key}: ${type};`;
  });

  const appThemeContent =
    customTypeEntries.length > 0
      ? customTypeEntries.join("\n")
      : "  // No custom properties found";
  const themeEntries = themeNames
    .map((name) => `  ${name}: AppTheme;`)
    .join("\n");

  return `
    export interface AppTheme extends DuckTheme {
${appThemeContent}
}

export type AppThemes = {
${themeEntries}
};
`;
}
