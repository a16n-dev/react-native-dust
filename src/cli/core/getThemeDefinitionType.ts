import {
  quicktype,
  InputData,
  jsonInputForTargetLanguage,
} from 'quicktype-core';
import { Config } from '../../config';

async function generateTypeInterfaceFromObjects(
  objs: any[],
  interfaceName: string
): Promise<string> {
  const jsonInput = jsonInputForTargetLanguage('typescript');
  await jsonInput.addSource({
    name: interfaceName,
    samples: objs.map((v) => JSON.stringify(v)),
  });

  const inputData = new InputData();
  inputData.addInput(jsonInput);

  const { lines } = await quicktype({
    inputData,
    lang: 'typescript',
    inferMaps: false,
    rendererOptions: {
      'just-types': true,
      'explicit-unions': true,
      'prefer-unions': false,
    },
  });

  return lines.join('\n');
}

/**
 * This takes in a theme object and generates a TypeScript type to account for any
 * custom properties the user may have defined in their theme.
 *
 * If the user has defined multiple themes, any properties that are not defined in all themes
 * will be marked as optional in the generated types
 */
export async function getThemeDefinition(config: Config): Promise<string> {
  const allThemeObjects = Object.values(config.themes);

  const appThemeInterface = await generateTypeInterfaceFromObjects(
    allThemeObjects,
    'AppTheme'
  );

  if (config.options?.mode === 'unistyles') {
    const themeNames = Object.keys(config.themes);

    const themeEntries = themeNames
      .map((name) => `  ${name}: AppTheme;`)
      .join('\n');

    const appThemesType = `
export declare const themes: {
${themeEntries}
};`;

    const appBreakpointsType = `
export declare const breakpoints: {
${Object.entries(config.breakpoints ?? {})
  .map(([key, value]) => `  ${key}: ${value};`)
  .join('\n')}
}`;

    return `${appThemeInterface}${appThemesType}\n${appBreakpointsType}`;
  }

  const appThemeType = `
export declare const theme: AppTheme;`;

  const appBreakpointsType = `
export declare const breakpoints: {
${Object.entries(config.breakpoints ?? {})
  .map(([key, value]) => `  ${key}: ${value};`)
  .join('\n')}
}`;

  return `${appThemeInterface}${appThemeType}\n${appBreakpointsType}`;
}
