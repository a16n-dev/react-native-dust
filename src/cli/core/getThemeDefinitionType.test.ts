import { expect, test } from 'vitest';
import { generateTypeInterfaceFromObjects } from './getThemeDefinitionType';

test('Generates a type inline correctly', async () => {
  const result = await generateTypeInterfaceFromObjects(
    [
      {
        colors: {
          blue: {
            light: '#0000FF',
            default: '#0000CD',
            dark: '#00008B',
          },
          red: {
            light: '#FFA07A',
            default: '#FF0000',
            dark: '#8B0000',
          },
        },
      },
    ],
    'Example'
  );

  console.log(result);

  expect(result).toMatchSnapshot();
});
