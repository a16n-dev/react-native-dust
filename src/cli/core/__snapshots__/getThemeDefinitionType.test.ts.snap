// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generates a type inline correctly 1`] = `
"export interface Example {
    colors: Colors;
}

export interface Colors {
    blue: Blue;
    red:  Blue;
}

export interface Blue {
    light:   string;
    default: string;
    dark:    string;
}
"
`;
